wrappers/services.go

```diff
@@ -375,120 +376,202 @@ func userDaemonReload() error {
 	return cli.ServicesDaemonReload(ctx)
 }
 
-// AddSnapServicesOptions is a struct for controlling the generated service
-// definition for a snap service.
-type AddSnapServicesOptions struct {
-	Preseeding              bool
-	VitalityRank            int
-	RequireMountedSnapdSnap bool
-}
-
-// AddSnapServices adds service units for the applications from the snap which
-// are services. The services do not get enabled or started.
-func AddSnapServices(s *snap.Info, opts *AddSnapServicesOptions, inter interacter) (err error) {
-	if s.Type() == snap.TypeSnapd {
-		return fmt.Errorf("internal error: adding explicit services for snapd snap is unexpected")
-	}
+// EnsureSnapServices will ensure that the specified snap services' file states
+// are up to date with the specified options and infos.
+func EnsureSnapServices(snaps map[*snap.Info]*AddSnapServicesOptions, inter interacter) (modified map[*snap.Info][]*snap.AppInfo, err error) {
+	modified = make(map[*snap.Info][]*snap.AppInfo)
 
-	if opts == nil {
-		opts = &AddSnapServicesOptions{}
-	}
-
-	// TODO: remove once services get enabled on start and not when created.
-	preseeding := opts.Preseeding
+	preseeding := false
 
 	// note, sysd is not used when preseeding
 	sysd := systemd.New(systemd.SystemMode, inter)
-	var written []string
-	var writtenSystem, writtenUser bool
+	modifiedUnits := make(map[string]*osutil.FileState)
```

Comment:
well it's a fair guess, but that assumes that there is only 1 service per snap that needs to be modified, but we don't really know which services will need to be modified or not, and we also don't know how many services there are in each snap. I would say if we are trying to save on memory allocation like this we should profile this to know that the memory allocation is problematic, otherwise we are just cargo culting pre-allocation of things, when in this case we really don't have a good estimate of how much memory we need. 

if you feel strongly, I can certainly add this, but it feels cargo culting like this to me given that we don't have a good estimate on the upper bound.